linchpin/provision/roles/libvirt/tasks/provision_libvirt_node.yml

```diff
@@ -23,69 +27,279 @@
+  when: uri_hostname =='localhost' and ssh_key_stat_local.stat.exists == false
+
+- name: "remote: Generate ssh keys when they donot exist"
+  shell: "ssh-keygen -t rsa -f {{ default_ssh_key_path | default('~/.ssh/') }}/{{ res_def['ssh_key'] | default(res_grp['resource_group_name']) }} -N '';cat {{ default_ssh_key_path | default('~/.ssh/') }}/{{ res_def['ssh_key'] }} >> ~/.ssh/authorized_keys"
+  args:
+    creates: "{{ default_ssh_key_path | default('~/.ssh/') }}/{{ res_def['ssh_key'] | default(res_grp['resource_group_name']) }}"
+  remote_user: "{{ res_def['remote_user'] | default('root') }}"
+  delegate_to: "{{ uri_hostname }}"
+  when: uri_hostname != 'localhost' and ssh_key_stat_remote.stat.exists == false
+
+- name: "local: cat out contents of public key"
+  shell: "cat {{ default_ssh_key_path | default('~/.ssh/') }}/{{ res_def['ssh_key'] | default(res_grp['resource_group_name']) }}.pub"
+  register: pubkey_local
+  when: uri_hostname == 'localhost'
+
+- name: "remote: cat out contents of public key"
+  shell: "cat {{ default_ssh_key_path | default('~/.ssh/') }}/{{ res_def['ssh_key'] | default(res_grp['resource_group_name']) }}.pub"
+  register: pubkey_remote
+  remote_user: "{{ res_def['remote_user'] | default('root') }}"
+  delegate_to: "{{ uri_hostname }}"
+  when: uri_hostname != 'localhost'
+
+- name: "Add additional storage"
+  command: "qemu-img resize {{ definition[1] }}/{{ definition[0] }}_{{ definition[4] }}.{{ definition[2] }} +{{ definition[3] }}"
+  with_nested:
+    - ["{{ libvirt_resource_name }}"]
+    - ["{{ libvirt_image_path }}"]
+    - ["{{ img_src_ext }}"]
+    - ["{{ res_def['additional_storage'] | default('1G') }}"]
+    - "{{ res_count.stdout }}"
   loop_control:
-    loop_var: net_item
+    loop_var: definition
+  become: yes
+  when: res_def['additional_storage'] is defined and node_exists['failed'] is defined and uri_hostname == 'localhost'
+
+- name: "Add additional storage"
+  command: "qemu-img resize {{ definition[1] }}/{{ definition[0] }}_{{ definition[4] }}.{{ definition[2] }} +{{ definition[3] }}"
+  with_nested:
+    - ["{{ libvirt_resource_name }}"]
+    - ["{{ libvirt_image_path }}"]
+    - ["{{ img_src_ext }}"]
+    - ["{{ res_def['additional_storage'] | default('1G') }}"]
+    - "{{ res_count.stdout }}"
+  loop_control:
+    loop_var: definition
+  become: yes
+  remote_user: "{{ res_def['remote_user'] | default('root') }}"
+  delegate_to: "{{ uri_hostname }}"
+  when: res_def['additional_storage'] is defined and node_exists['failed'] is defined and uri_hostname != 'localhost'
+
+- name: "Create directories"
+  file:
+    path: "/tmp/vm-{{ definition[0] }}_{{ definition[1] }}"
+    state: "directory"
+  with_nested:
+    - ["{{ libvirt_resource_name }}"]
+    - "{{ res_count.stdout }}"
+  loop_control:
+    loop_var: definition
+  remote_user: "{{ res_def['remote_user'] | default('root') }}"
```

Comment:
This seems to be a theme. I think we should take a minute and discuss what this is doing and why root is necessary. It may matter, but let's discuss and make sure it's the best way to do things. There may be better ways to grant remote (or local) privilege escalation.

Reply:
'root' is not necessary , its defaulted . I will change it current;y active user running linchpin . in case of remote_machine. As its tmp folder , it doesnt need privilege acceleration in this case